File: Exercism/python/isbn-verifier/isbn_verifier.py
from itertools import count
from operator import mul


def verify(isbn):
    try:
        control = 10 if isbn[-1] == "X" else int(isbn[-1])
    except (ValueError, IndexError):
        return False
    counter = count(10, -1)
    checksum = sum(
        map(
            mul,
            (int(d) for d in isbn[:-1] if d.isdigit()),
            counter,
        )
    ) + control
    if next(counter) != 1:
        return False
    return checksum % 11 == 0

File: Exercism/haskell/isbn-verifier/src/IsbnVerifier.hs
module IsbnVerifier (isbn) where

import qualified Data.Char as C
import Data.Monoid

isbn :: String -> Bool
isbn [] = False
isbn xs = ifAllTrue isbnChecks
  where
    isbnChecks = [all (\x -> C.isDigit x || x == '-') $ init xs
                 ,length digits == 9
                 ,C.isDigit (last xs) || C.toLower (last xs) == 'x'
                 ,checkSum `mod` 11 == 0
                 ]
    digits  = map C.digitToInt $ filter C.isDigit $ init xs
    control = (\x -> if C.isDigit x then C.digitToInt x else 10) $ last xs
    checkSum = sum $ zipWith (*) [1..10] (control:reverse digits)

ifAllTrue :: [Bool] -> Bool
ifAllTrue = getAll . mconcat . map All
